detector.py

```diff
@@ -49,30 +55,232 @@ def detect(self, image, camera: Camera) -> List[PersonView]:
+        for i, image in enumerate(images):
+            img_name = 'image{}.png'.format(i)
+            # OpenPose binary reads images from a given directory, so we need to write images to the directory first
+            cv2.imwrite(os.path.join(self.images_folder, img_name), image)
+
+        # run detection
+        p = subprocess.Popen(self.cmd, cwd=self.binary_home, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+        cmd_result = p.communicate()
+        if p.returncode != 0:
+            logger.error("OpenPose binary run failed!\nSTDOUT: {}\nSTDERR: {}".format(cmd_result[0], cmd_result[1]))
+            raise RuntimeError("OpenPose binary run failed!")
+        else:
+            logger.debug("OpenPose binary run success. STDOUT: {}".format(cmd_result[0]))
+
+        # parse detection results to person views
+        for i, image in enumerate(images):
+            result_name = 'image{}_keypoints.json'.format(i)
+            views = self.load_valid_persons_from_json(os.path.join(self.results_folder, result_name), image, cameras[i])
+            results.append(views)
+
+        # TODO delete tmp images
+
+        return results
+
+    def load_valid_persons_from_json(self, json_path, image, camera: Camera) -> List[PersonView]:
+        with open(json_path) as json_file:
+            detection_result = json.load(json_file)
+
+        results = []
+        for person in detection_result['people']:
+            body_parts = self.get_body_parts_from_keypoints(person['pose_keypoints_2d'])
+            person_image = self.get_person_subimage(image, body_parts)
+            neck, hip_center = self.get_neck_and_hip_coordinates(body_parts)
+            if not neck or not hip_center:
+                logger.warning("Person does not have nose or hips detected.")
+                continue
+
+            results.append(PersonView(image, person_image, camera, neck, hip_center))
+
+        return results
+
+    @staticmethod
+    def get_body_parts_from_keypoints(keypoints: List[float], part_confidence_threshold=0.7) -> List[Optional[Tuple[int, int]]]:
+        """
+        Body parts are loaded only if detection confidence (0 to 1) is higher than the part confidence threshold.
+        :return: body parts coordinates (x, y); order of parts can be found at:
+        https://github.com/CMU-Perceptual-Computing-Lab/openpose/blob/master/doc/output.md#keypoint-ordering
+        """
+        body_parts = []
+        i = 2  # starting from the first confidence index
+        for detection_confidence in keypoints[2::3]:
+            # keypoint is defined as (part X, part Y, probability)
+            if detection_confidence >= part_confidence_threshold:
+                body_parts.append((int(keypoints[i-2]), int(keypoints[i-1])))
+            else:
+                body_parts.append(None)
+
+            i += 3  # iterating by 3
+
+        return body_parts
+
+    @staticmethod
+    def get_person_subimage(image, body_parts: List[Optional[Tuple[int, int]]]):
```

Comment:
How does this differ from `person.PersonView.get_torso_subimage()`?

Reply:
This is a method to get image of the whole person (for all body parts). It is in PovaPose detector, so I preserved the functionality. This could be combined with background substraction to get image of just a body.